src/components/SEO/index.tsx

```diff
@@ -0,0 +1,119 @@
+import path from 'path'
+import React, { FC } from 'react'
+import Helmet from 'react-helmet'
+import getSharingImage from '@jlengstorf/get-share-image'
+import { graphql, PageProps } from 'gatsby'
+import SchemaOrg from './SchemaOrg'
+import config from '../../../config/website'
+import { useStaticQuery } from 'gatsby'
+import { Maybe, PostQuery, SeoQuery } from '../../types/generated'
+import { Defined } from '../../types/Defined'
+
+interface FrontmatterType {
+  title?: string;
+  description?: string;
+  tagline?: string;
+  slug?: string;
+  datePublished?: boolean;
+}
+
+interface SeoPropsType {
+  isBlogPost?: boolean;
+  postData?: {
+    childMarkdownRemark?: {
+      frontmatter?: FrontmatterType;
+    }
+  }
+  frontmatter?: Maybe<Defined<PageProps<PostQuery>['data']['mdx']>['frontmatter']>
+  postImage?: string | null;
+}
+
+const SEO: FC<SeoPropsType> = ({ postData, frontmatter, isBlogPost }) => {
+  const { site } = useStaticQuery<SeoQuery>(graphql`
+    query SEO {
+      site {
+        siteMetadata {
+          title
+          description
+          canonicalUrl
+          image
+          author {
+            name
+          }
+          organization {
+            name
+            url
+            logo
+          }
+          social {
+            twitter
+            fbAppID
+          }
+        }
+      }
+    }
+  `);
+
+  const seo = site?.siteMetadata ?? {};
+  const postMeta = frontmatter ?? postData?.childMarkdownRemark?.frontmatter ?? {}
+  const title = isBlogPost ? postMeta.title ?? '' : config.siteTitle;
+  const tagline = postMeta.tagline || 'Read more'
+  const description = postMeta.description || seo.description
+  const url = postMeta.slug
+    ? `${seo.canonicalUrl}${path.sep}${postMeta.slug}`
+    : seo.canonicalUrl
+  // @ts-expect-error datePublished is not part of postMeta
+  const datePublished = isBlogPost ? postMeta?.datePublished ?? false : false
+
+  const socialImage = getSharingImage({
+    title,
+    tagline,
+    cloudName: config.cloudinaryCloudName,
+    imagePublicID: config.cloudinaryImagePublicID,
+    titleFont: 'Roboto',
+    titleExtraConfig: '_bold',
+    taglineFont: 'Roboto',
+  })
+
+  const image = isBlogPost ? socialImage : seo.image
+
+  return (
+    <React.Fragment>
+      <Helmet>
+        {/* General tags */}
+        <title>{title}</title>
+        {description && <meta name="description" content={description} />}
```

Comment:
all the properties now have optional value, so in order to render it without `undefined` I check it beforehand

Reply:
sounds good! 💯